libhif/hif-source.c

```diff
@@ -1634,33 +1634,63 @@ hif_source_copy_package (HyPackage pkg,
 			    hif_source_copy_progress_cb, state, error);
 }
 
-struct DownloadState {
-	HifState *state;
+typedef struct
+{
 	gchar *last_mirror_url;
 	gchar *last_mirror_failure_message;
-};
+} MirrorFailureData;
+
+typedef struct
+{
+	HyPackage pkg;
+	HifState *state;
+	MirrorFailureData *mirror_failure_data;
+} PackageDownloadData;
 
 static int
 package_download_update_state_cb (void *user_data,
 				  gdouble total_to_download,
 				  gdouble now_downloaded)
 {
-	struct DownloadState *dlstate = user_data;
-	return hif_source_update_state_cb (dlstate->state, total_to_download, now_downloaded);
+	PackageDownloadData *data = user_data;
+	HifState *state_child;
+
+	state_child = hif_state_get_child (data->state);
```

Comment:
Oh hm, you are totally right. Let me poke at this a bit, a bit confused how this even works at all right now without completely blowing up...